File: java/selenium/src/test/java/dev/selenium/request/ModifyHeaderTest.java
package dev.selenium.request;

import dev.selenium.TestBase;
import java.net.URI;
import java.util.function.Predicate;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.firefox.FirefoxDriver;

public class ModifyHeaderTest extends TestBase {

  @Test
  public void addHeader() {
    Predicate<URI> filter = uri -> uri.getHost().equals("httpbin.org");

    ((FirefoxDriver) driver)
        .network()
        .addRequestHandler(filter, req -> req.addHeader("X-Test", "true"));

    // .  RequestInterceptOptions options =
    // RequestInterceptOptionsBuilder.addFilter(Map.of("host", "httpbin.org").build();
    //    driver
    //        .network()
    //        .addRequestHandler(options, req -> req.addHeader("X-Test", "true")));

    driver.get("https://httpbin.org/headers");
    Assertions.assertEquals(
        "X-Test \"true\"", driver.findElement(By.id("/headers/X-Test")).getText());
  }

  @Test
  public void removeHeader() {
    Predicate<URI> filter = uri -> uri.getHost().equals("httpbin.org");

    ((FirefoxDriver) driver)
        .network()
        .addRequestHandler(filter, req -> req.removeHeader("upgrade-insecure-requests"));

    // .  RequestInterceptOptions options =
    // RequestInterceptOptionsBuilder.addFilter(Map.of("host", "httpbin.org").build();
    //    driver
    //        .network()
    //        .addRequestHandler(options, req-> req.removeHeader("upgrade-insecure-requests"));

    driver.get("https://httpbin.org/headers");

    Assertions.assertTrue(
        driver.findElements(By.id("/headers/Upgrade-Insecure-Requests")).isEmpty());
  }
}

File: dotnet/Selenium/Request/ModifyHeaderTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.BiDi.Network;

namespace Selenium.Request;

[TestClass]
public class ModifyHeaderTest : TestBase
{
    [TestMethod]
    public async Task AddHeader()
    {
        await using var intercept = await Bidi.Network.InterceptRequestAsync(async req =>
        {
            var headers = new List<Header>(req.Request.Headers) { new("X-Test", "true") };

            await req.ContinueAsync(new ContinueRequestOptions { Headers = headers });
        });

        // var options = new RequestInterceptOptions {
        //               UrlPatterns = new List<UrlPattern> { new UrlPattern("https://httbin.org/") },
        //           };
        //
        // await using var registration = await driver.Network.AddRequestHandlerAsync(
        //     options,
        //     req =>
        //     {
        //          req.AddHeader("X-Test", "true");
        //      }
        // );

        await Driver.Navigate().GoToUrlAsync("https://httpbin.org/headers");
        Assert.AreEqual("X-Test \"true\"", Driver.FindElement(By.Id("/headers/X-Test")).Text);
    }

    [TestMethod]
    public async Task RemoveHeader()
    {
        await using var intercept = await Bidi.Network.InterceptRequestAsync(async req =>
        {
            var headers = new List<Header?>(req.Request.Headers);
            var remove = headers.Find(h => h?.Name == "Upgrade-Insecure-Requests");
            headers.Remove(remove);

            await req.ContinueAsync(new ContinueRequestOptions { Headers = headers! });
        });

        
        // await using var registration = await driver.Network.AddRequestHandlerAsync(
        //     options,
        //     req =>
        //     {
        //          req.removeHeader("Upgrade-Insecure-Requests")
        //      }
        // );

        await Driver.Navigate().GoToUrlAsync("https://httpbin.org/headers");

        Assert.AreEqual(0, Driver.FindElements(By.Id("/headers/Upgrade-Insecure-Requests")).Count);
    }
}